File: go/test/utility_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/whois-domain-monitoring-sdk/go"
	"github.com/voxgig-sdk/whois-domain-monitoring-sdk/go/core"

	vs "github.com/voxgig-sdk/whois-domain-monitoring-sdk/go/utility/struct"
)

func TestUtilityEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Utility(nil)
		if ent == nil {
			t.Fatal("expected non-nil UtilityEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := utilityBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "utility." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set WHOIS_DOMAIN_MONITORING_TEST_UTILITY_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		utilityRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.utility", setup.data)))
		var utilityRef01Data map[string]any
		if len(utilityRef01DataRaw) > 0 {
			utilityRef01Data = core.ToMapAny(utilityRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = utilityRef01Data

		// LOAD
		utilityRef01Ent := client.Utility(nil)
		utilityRef01MatchDt0 := map[string]any{}
		utilityRef01DataDt0Loaded, err := utilityRef01Ent.Load(utilityRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if utilityRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

	})
}

func utilityBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "utility", "UtilityTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read utility test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse utility test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"utility01", "utility02", "utility03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("WHOIS_DOMAIN_MONITORING_TEST_UTILITY_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"WHOIS_DOMAIN_MONITORING_TEST_UTILITY_ENTID": idmap,
		"WHOIS_DOMAIN_MONITORING_TEST_LIVE":      "FALSE",
		"WHOIS_DOMAIN_MONITORING_TEST_EXPLAIN":   "FALSE",
		"WHOIS_DOMAIN_MONITORING_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["WHOIS_DOMAIN_MONITORING_TEST_UTILITY_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["WHOIS_DOMAIN_MONITORING_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["WHOIS_DOMAIN_MONITORING_APIKEY"],
			},
			extra,
		})
		client = sdk.NewWhoisDomainMonitoringSDK(core.ToMapAny(mergedOpts))
	}

	live := env["WHOIS_DOMAIN_MONITORING_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["WHOIS_DOMAIN_MONITORING_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/UtilityEntityTest.php
<?php
declare(strict_types=1);

// Utility entity test

require_once __DIR__ . '/../whoisdomainmonitoring_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class UtilityEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = WhoisDomainMonitoringSDK::test(null, null);
        $ent = $testsdk->Utility(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = utility_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["load"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "utility." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set WHOIS_DOMAIN_MONITORING_TEST_UTILITY_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $utility_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.utility")));
        $utility_ref01_data = null;
        if (count($utility_ref01_data_raw) > 0) {
            $utility_ref01_data = Helpers::to_map($utility_ref01_data_raw[0][1]);
        }

        // LOAD
        $utility_ref01_ent = $client->Utility(null);
        $utility_ref01_match_dt0 = [];
        $utility_ref01_data_dt0_loaded = $utility_ref01_ent->load($utility_ref01_match_dt0, null);
        $this->assertNotNull($utility_ref01_data_dt0_loaded);

    }
}

function utility_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/utility/UtilityTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = WhoisDomainMonitoringSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["utility01", "utility02", "utility03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("WHOIS_DOMAIN_MONITORING_TEST_UTILITY_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "WHOIS_DOMAIN_MONITORING_TEST_UTILITY_ENTID" => $idmap,
        "WHOIS_DOMAIN_MONITORING_TEST_LIVE" => "FALSE",
        "WHOIS_DOMAIN_MONITORING_TEST_EXPLAIN" => "FALSE",
        "WHOIS_DOMAIN_MONITORING_APIKEY" => "",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["WHOIS_DOMAIN_MONITORING_TEST_UTILITY_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["WHOIS_DOMAIN_MONITORING_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            // FIRST, so the generated fields below win: sdk-test-control.json's
            // test.client.options adds to the live client, it does not redirect it.
            Runner::live_client_options(),
            [
                "apikey" => $env["WHOIS_DOMAIN_MONITORING_APIKEY"],
            ],
            // ismap, not a plain "?? []" default: an empty PHP array is a
            // LIST, and a non-map later entry REPLACES the accumulated map in
            // merge - so the no-extras call discarded live_client_options()
            // and the apikey/server map above it.
            Vs::ismap($extra) ? $extra : new \stdClass(),
        ]);
        $client = new WhoisDomainMonitoringSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["WHOIS_DOMAIN_MONITORING_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["WHOIS_DOMAIN_MONITORING_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
